app/scripts/lib/transaction-metrics.ts

```diff
@@ -0,0 +1,879 @@
+      controllerActions,
+      transactionMeta,
+      payload: {
+        properties,
+        sensitiveProperties,
+      },
+    });
+
+    finalizeTransactionEventFragment({
+      eventName: TransactionMetaMetricsEvent.approved,
+      controllerActions,
+      transactionMeta,
+    });
+  };
+
+export const onTransactionFinalized =
+  ({ controllerActions, provider }: TrackerDependencyMap) =>
+  async ({
+    transactionMeta,
+    error,
+  }: {
+    transactionMeta: TransactionMeta;
+    error: string;
+  }) => {
+    if (!transactionMeta) {
+      return;
+    }
+
+    const extraParams = {} as Record<string, any>;
+    if (error) {
+      // This is a failed transaction
+      extraParams.error = error;
+    } else {
+      const { txReceipt } = transactionMeta;
+
+      extraParams.gas_used = txReceipt.gasUsed;
+
+      const { submittedTime } = transactionMeta;
+
+      if (submittedTime) {
+        extraParams.completion_time =
+          getTransactionCompletionTime(submittedTime);
+      }
+
+      if (txReceipt.status === '0x0') {
+        extraParams.status = METRICS_STATUS_FAILED;
+      }
+    }
+
+    const { properties, sensitiveProperties } =
+      await buildEventFragmentProperties({
+        transactionMeta,
+        controllerActions,
+        provider,
+        extraParams,
+      });
+
+    createTransactionEventFragment({
+      eventName: TransactionMetaMetricsEvent.finalized,
+      controllerActions,
+      transactionMeta,
+      payload: {
+        properties,
+        sensitiveProperties,
+      },
+    });
+
+    updateTransactionEventFragment({
+      eventName: TransactionMetaMetricsEvent.finalized,
+      controllerActions,
+      transactionMeta,
+      payload: {
+        properties,
+        sensitiveProperties,
+      },
+    });
+
+    finalizeTransactionEventFragment({
+      eventName: TransactionMetaMetricsEvent.finalized,
+      controllerActions,
+      transactionMeta,
+    });
+  };
+
+export const onTransactionDropped =
+  ({ controllerActions, provider }: TrackerDependencyMap) =>
+  async ({ transactionMeta }: { transactionMeta: TransactionMeta }) => {
+    if (!transactionMeta) {
+      return;
+    }
+
+    const extraParams = {
+      dropped: true,
+    };
+
+    const { properties, sensitiveProperties } =
+      await buildEventFragmentProperties({
+        transactionMeta,
+        controllerActions,
+        provider,
+        extraParams,
+      });
+
+    createTransactionEventFragment({
+      eventName: TransactionMetaMetricsEvent.finalized,
+      controllerActions,
+      transactionMeta,
+      payload: {
+        properties,
+        sensitiveProperties,
+      },
+    });
+
+    updateTransactionEventFragment({
+      eventName: TransactionMetaMetricsEvent.finalized,
+      controllerActions,
+      transactionMeta,
+      payload: {
+        properties,
+        sensitiveProperties,
+      },
+    });
+
+    finalizeTransactionEventFragment({
```

Comment:
As this pattern is repeated a few times with what appear to be the same arguments, is it worth keeping these separate methods but also creating a wrapper method to call create, update, and finalize, such as `createUpdateFinalizeTransactionEventFragment`?

Reply:
We could even include the call to `buildEventFragmentProperties` as all these exported functions seem to have very similar logic.

Reply:
We have now `createUpdateFinalizeTransactionEventFragment` responsible to do all actions, done in https://github.com/MetaMask/metamask-extension/pull/21330/commits/a5a32f8ed27f3f3fd223c3e07fa5c81b011bd409